components/layout_2020/flow/inline/construct.rs

```diff
@@ -253,18 +274,30 @@ impl InlineFormattingContextBuilder {
 
         let new_range = self.current_text_offset..self.current_text_offset + new_text.len();
         self.current_text_offset = new_range.end;
+        let new_character_range =
+            self.current_character_offset..self.current_character_offset + new_text.chars().count();
+        self.current_character_offset = new_character_range.end;
         self.text_segments.push(new_text);
 
         if let Some(inline_item) = self.inline_items.last() {
             if let InlineItem::TextRun(text_run) = &mut *inline_item.borrow_mut() {
                 text_run.borrow_mut().text_range.end = new_range.end;
+                text_run.borrow_mut().character_range.end = new_character_range.end;
                 return;
             }
         }
 
         self.inline_items
             .push(ArcRefCell::new(InlineItem::TextRun(ArcRefCell::new(
-                TextRun::new(info.into(), info.style.clone(), new_range),
+                TextRun::new(
+                    info.into(),
+                    info.style.clone(),
+                    new_range,
+                    new_character_range,
+                    insertion_point,
```

Comment:
Instead of sending the starting position, why not just adjust the ranges sent the the constructor?